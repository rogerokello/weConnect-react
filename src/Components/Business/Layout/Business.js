import React, { Component } from 'react';

export const Business = ({header,fetchDataLoader, form}) => {
    return(
        <div className="container-fluid text-center">
            {fetchDataLoader}
    
            <div className="row content">
                <div className="col-sm-3">
                </div>
                <div className="col-sm-6">
                    <h2 className="text-center">{header}</h2>
                    {form}   
                </div>
                <div className="col-sm-3">
                </div>
            </div>
        </div>
    );
}
