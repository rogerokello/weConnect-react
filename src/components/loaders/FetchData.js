import React, { Component } from 'react';
import Loader from 'react-loader-spinner';

export const FetchData = ({pendingTask, style}) => {
    return (
        <div style={style}>
            {pendingTask?
                <span>
                    <Loader type="ThreeDots"
                            color="#00BFFF"
                            height="100"	
                            width="100" 
                    />
                </span>
                :
                <span></span>
            }
        </div>
    )
}