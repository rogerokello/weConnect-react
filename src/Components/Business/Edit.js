import React, { Component } from "react";
import {withRouter} from "react-router-dom";
import {NotificationManager} from 'react-notifications';
import {connect} from 'react-redux';
import {PropTypes} from "prop-types";
import Loader from 'react-loader-spinner';

import store from "../../store";
import {loaderPosition} from "../Styles/loaderPosition";
import Navbar from "../Layout/Common/NavigationBar";
import {clearEditMessageStatus} from "../../Actions/actionCreators";
import {editBusiness} from "../../Actions/businessActions";

import {Footer} from "../../Components/Layout/Common/Footer";

import {FetchData} from "../Loaders/FetchData";

import {Business} from "./Layout/Business";

import {Form} from "./Form";


export class Editbusiness extends Component {
	constructor(props) {
        super(props);
		this.submitForm = this.submitForm.bind(this);
		this.state = {
			currentBusiness:{}
		};
    }

	componentWillReceiveProps(recieved){
		if(recieved && recieved.editMessage.status==="success"){
			this.props.history.push("/businesslist");
			NotificationManager.info(recieved.editMessage.message, "", 5000);
			store.dispatch(clearEditMessageStatus());
		}

		if(recieved && recieved.editMessage.status === "failure"){
			NotificationManager.info(recieved.editMessage.message, "", 5000);
			store.dispatch(clearEditMessageStatus());
		}

	};


    componentWillMount(){
        if(localStorage.getItem("access_token") === null){
            this.props.history.push("/login");
        }
	}

	componentDidMount(){
		this.setState({currentBusiness:this.props.currentBusiness});
	}

	handleEdit = (e) =>{
		let field = e.target.name;
		let business = this.state.currentBusiness;

		if(e.target.type !== "hidden"){
			business[field] = e.target.value;
		}

		this.setState({
			currentBusiness:business
		});
	}
	
	submitForm = (e) => {
		e.preventDefault();

		if(this.state.currentBusiness){
			let business = this.state.currentBusiness;
			let businesslist = {};
			
			//Remove fields with number
			for (var key in business) {
				if (business.hasOwnProperty(key)) {

					if((typeof business[key]) === "number"){
						continue;
					}
					businesslist[key] = business[key];
				}
			}

			this.props.editBusiness(
				JSON.stringify(businesslist),
				this.state.currentBusiness.id
			);
		}
	}

	render() {

		return (
			
			<div>
				
				<Navbar/>


				<Business
					header={"Edit business"}
					fetchDataLoader={<FetchData style={loaderPosition} pendingTask={this.props.pendingTask} />}
					form={
						<Form	id="editBusinessForm" 
								submitForm={this.submitForm.bind(this)}
								handleEdit={this.handleEdit.bind(this)}
								currentBusiness={this.props.currentBusiness}
								formType={"Edit"} 
						/>
					}
				/>	
				<Footer/>
			</div>
		);
	}
}

// Find current business based on ID passed in URL
function findCurrentBusiness(businesses, id = -1) {

    // Find business for given id
    return businesses.find(business => {
        return parseInt(business.id, 10) === parseInt(id, 10);
    });
}

Editbusiness.propTypes = {
	editBusiness: PropTypes.func.isRequired,
	business:PropTypes.object
}

const mapStateToProps = (state, ownProps) => {
	let currentBusiness;
	if(state){
		// currentBusiness = state.business.businesses.message.length ? findCurrentBusiness(state.business.businesses.message, ownProps.match.params.id) : null;
		if(state.business.businesses.message.length) {
			currentBusiness = findCurrentBusiness(state.business.businesses.message, ownProps.match.params.id);
		}else{
			currentBusiness = null;
		}
	}
	return {
		currentBusiness,
		business: state.business.businesses,
		editMessage: state.business.editbusinessMessage,
		pendingTask: state.pendingTasksReducer,
    }
};


export default withRouter(connect(mapStateToProps, {editBusiness})(Editbusiness));