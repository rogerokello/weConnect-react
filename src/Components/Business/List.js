import React, { Component } from "react";
import {withRouter, NavLink} from "react-router-dom";
import {NotificationManager} from 'react-notifications';
import {connect} from 'react-redux';
import {PropTypes} from "prop-types";
import Loader from 'react-loader-spinner';

import Navbar from "../Layout/Common/NavigationBar";
import Pagination from "../Business/Layout/Pagination";
import store from "../../store";
import {getAllBusiness, deleteBusiness} from "../../Actions/businessActions";
import {clearDeleteMessageStatus} from "../../Actions/actionCreators";

import {Footer} from "../../Components/Layout/Common/Footer";

import {FetchData} from "../Loaders/FetchData";
import {generateBusinessesByPage} from "../../Helpers/generateBusinessesByPage"

import {ListRows} from "./Layout/ListRows";
import {TableList} from "./Layout/TableList";

import {
	viewbuttonStyle,
	editbuttonStyle,
	loaderPosition
} from "./Layout/Styles/styles";




export class Businesslist extends Component {

	
	componentWillReceiveProps(recieved){
		console.log("Component Recieved Props")
		if(recieved && recieved.deletemessage.status==="success"){
			NotificationManager.success("Business Deleted successfuly", "", 5000);
			store.dispatch(clearDeleteMessageStatus());
		}

		if(recieved && recieved.business.status === "failure"){
			NotificationManager.info(recieved.business.message, "", 5000);
		}

		if(store.getState().connection.tokenExpired===true){
			NotificationManager.success("Please login again to refresh your credentials", "", 5000);
			this.props.history.push("/login");
		}

	}


  	componentDidMount(){
		if(localStorage.getItem("access_token") === null){
			this.props.history.push("/login");
		}
		
		//Get all the businesses
		this.props.getAllBusiness();

  	}


  	deleteBusinessRemote(id){

		const popup = window.confirm('Are you sure you want to delete this business?'); 
		
		if (popup === true) {
			this.props.deleteBusiness(id);
		}
	}

	render() {
		
		
		//Spread to extract all businesses from props
		//into a dictionary
		const businesses = Object.values({...this.props.business});


		if(businesses.length > 0){

			// Make an inplace arrangement of the elements.
			// The last one added should become first
			Array.prototype.reverse.call(businesses);

		}
		
		return (
			<div id="tableList">
				<Navbar {...this.props}/>
					
				<TableList
					loadingStatus={<FetchData pendingTask={this.props.pendingTask} style={loaderPosition} />}
					listRows={<ListRows 
						viewbuttonStyle={viewbuttonStyle}
						editbuttonStyle={editbuttonStyle}
						businesses={businesses}
						deleteRemoteBusiness={this.deleteBusinessRemote.bind(this)}
					/>}
					header={"Current list of Ugandan Businesses"}
				/>
					
				{ /* show pagination if there are more than 1 page */
                    this.props.pages > 1 && <Pagination has_next={this.props.has_next} has_prev={this.props.has_prev} pages={this.props.pages} currentPage={this.props.currentPage} getBusiness={this.props.getAllBusiness}/>
				}

				<Footer />
			</div>
		);
	}
}

Businesslist.propTypes = {
	deletemessage: PropTypes.object,
	business: PropTypes.array,
}



const mapStateToProps = (state, ownProps) => {
	console.log(state)
	// Set page number to 1 if no number in url params
	let pageNo = ownProps.match.params.pageNo || 1;
	let businesses = generateBusinessesByPage(state.business.businesses.message, pageNo);
	let pages;

	let has_next="";
	let has_prev="";
	let total_pages="";
	let total_number_of_items="";
	let current_page=""

	if(state.business.businesses){
		has_next = state.business.businesses.has_next;
		has_prev = state.business.businesses.has_prev;
		total_pages = state.business.businesses.total_pages;
		total_number_of_items = state.business.businesses.total_number_of_items;
		current_page = state.business.businesses.current_page;
	}

	if (state.business.businesses && state.business.businesses.message) {
		if (parseInt(total_number_of_items, 10) === NaN){
			pages = 1
		}else{
			pages = Math.ceil(parseInt(total_number_of_items, 10) / 6);
		}
    }else{
		pages = 1;
	}
	
    return {
		business: businesses,
		deletemessage: state.business.deletebusinessMessage,
		loading: state.business.loadingbusiness,
		user_id: state.user.signInMessage.id,
		pages: pages, // Determine number of pages for pagination
		currentPage: pageNo,
		tokenExpired: state.connection.tokenExpired,
		pendingTask: state.pendingTasksReducer,
		allbusinessinformation: state.business.businesses,
		has_next: has_next,
		has_prev: has_prev,
		total_pages,
		total_number_of_items,
		current_page,
    }
};


export default withRouter(connect(mapStateToProps, {getAllBusiness, deleteBusiness})(Businesslist));
