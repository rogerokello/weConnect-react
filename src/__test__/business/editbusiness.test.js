import React from 'react';
import expect from 'expect';
import {MemoryRouter} from 'react-router-dom';
import {shallow ,mount} from 'enzyme';
import configureStore from 'redux-mock-store';
import {Provider} from "react-redux";

import {FetchData} from "../../Components/Loaders/FetchData";


import mainReducer from "../../Reducers";
import {Editbusiness} from '../../Components/Business/Edit';

const middlewares = []
const mockStore = configureStore(mainReducer,middlewares)

const localStorageMock = {
    getItem: jest.fn(),
    setItem: jest.fn(),
    clear: jest.fn()
};

global.localStorage = localStorageMock;

const currentBusinessInfo = 
    {
        id: "1",
        location: "lira",
        name: "xedrox",
        category: "it"
    }

const initialState = {
    
}

const editMessage = {
    message: "",
    status: ""
}

const store = mockStore(initialState);
const mockEditbusinessfn = jest.fn();
const mockChangefn = jest.fn();

describe ('Edit business Page', () => {
    let wrapper;
    beforeEach( () => {
        const toBeMounted = <Provider store={store}>
                                <MemoryRouter>
                                    <Editbusiness 
                                        currentBusiness={currentBusinessInfo}
                                        editBusiness={mockEditbusinessfn} 
                                        editMessage={editMessage}   
                                    />
                                </MemoryRouter>
                            </Provider>
        
        wrapper = mount(toBeMounted);
    });
    it('checks edit form submits', () => {  
        expect(wrapper.find(Editbusiness).length).toEqual(1);
        expect(wrapper.find(FetchData).length).toEqual(1);
        wrapper.find('#editBusinessForm').at(1).simulate(
            'submit', 
            {preventDefault() {}}
        )
          
        expect(mockEditbusinessfn.mock.calls.length).toBe(1)
    });
    it('checks edit business component can edit', () => {
        let businessName = wrapper.find('#businessname');
        let category = wrapper.find('#businesscategory');
        let location = wrapper.find('#businesslocation');
        businessName.instance().value = 'jsjs';
        category.instance().value = "fdsafajsjs";
        location.instance().value = "jkl"
        businessName.instance().value = "fdsjsjs";
        expect(businessName.instance().value).toEqual('fdsjsjs');
        expect(category.instance().value).toEqual('fdsafajsjs');
        expect(location.instance().value).toEqual('jkl');        
    })
    it('matches snapshot',()=>{
        let wrapper = shallow(<Editbusiness currentBusiness={currentBusinessInfo}
            editBusiness={mockEditbusinessfn} 
            editMessage={editMessage} />) 
        expect(wrapper).toMatchSnapshot()
    })
    
    
});