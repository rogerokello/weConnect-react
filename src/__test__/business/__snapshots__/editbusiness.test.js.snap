// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit business Page matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Editbusiness
    currentBusiness={
        Object {
            "category": "it",
            "id": "1",
            "location": "lira",
            "name": "xedrox",
          }
    }
    editBusiness={[Function]}
    editMessage={
        Object {
            "message": "",
            "status": "",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <withRouter(Connect(Navbar)) />,
        <Unknown
          fetchDataLoader={
                    <Unknown
                              pendingTask={undefined}
                              style={
                                        Object {
                                                  "position": "absolute",
                                                  "right": "40%",
                                                  "top": "10%",
                                                  "width": "100%",
                                                }
                              }
                    />
          }
          form={
                    <Unknown
                              currentBusiness={
                                        Object {
                                                  "category": "it",
                                                  "id": "1",
                                                  "location": "lira",
                                                  "name": "xedrox",
                                                }
                              }
                              formType="Edit"
                              handleEdit={[Function]}
                              id="editBusinessForm"
                              submitForm={[Function]}
                    />
          }
          header="Edit business"
/>,
        <Unknown />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fetchDataLoader": <Unknown
            pendingTask={undefined}
            style={
                        Object {
                                    "position": "absolute",
                                    "right": "40%",
                                    "top": "10%",
                                    "width": "100%",
                                  }
            }
/>,
          "form": <Unknown
            currentBusiness={
                        Object {
                                    "category": "it",
                                    "id": "1",
                                    "location": "lira",
                                    "name": "xedrox",
                                  }
            }
            formType="Edit"
            handleEdit={[Function]}
            id="editBusinessForm"
            submitForm={[Function]}
/>,
          "header": "Edit business",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <withRouter(Connect(Navbar)) />,
          <Unknown
            fetchDataLoader={
                        <Unknown
                                    pendingTask={undefined}
                                    style={
                                                Object {
                                                            "position": "absolute",
                                                            "right": "40%",
                                                            "top": "10%",
                                                            "width": "100%",
                                                          }
                                    }
                        />
            }
            form={
                        <Unknown
                                    currentBusiness={
                                                Object {
                                                            "category": "it",
                                                            "id": "1",
                                                            "location": "lira",
                                                            "name": "xedrox",
                                                          }
                                    }
                                    formType="Edit"
                                    handleEdit={[Function]}
                                    id="editBusinessForm"
                                    submitForm={[Function]}
                        />
            }
            header="Edit business"
/>,
          <Unknown />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "fetchDataLoader": <Unknown
              pendingTask={undefined}
              style={
                            Object {
                                          "position": "absolute",
                                          "right": "40%",
                                          "top": "10%",
                                          "width": "100%",
                                        }
              }
/>,
            "form": <Unknown
              currentBusiness={
                            Object {
                                          "category": "it",
                                          "id": "1",
                                          "location": "lira",
                                          "name": "xedrox",
                                        }
              }
              formType="Edit"
              handleEdit={[Function]}
              id="editBusinessForm"
              submitForm={[Function]}
/>,
            "header": "Edit business",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
